File: nQueen.java
// Time Complexity : O(n!)
// Space Complexity : O(n^2)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

import java.util.*;

class Main {
    private static List<List<String>> result;

    public static List<List<String>> solveNQueens(int n) {
        result = new ArrayList<>();
        // null case
        if (n == 0)
            return result;
        // to place queen at particular index I am using boolean matrix of n*n
        boolean[][] box = new boolean[n][n];
        // calling backTrack function
        backTrack(box, 0, n);
        return result;

    }

    private static void backTrack(boolean[][] box, int r, int n) {
        // base case
        if (r == n) {
            // make a list of queen location at each row
            List<String> li = new ArrayList<>();

            for (int i = 0; i < n; i++) {
                // using string builder because string is immutable structure
                StringBuilder sb = new StringBuilder();
                for (int j = 0; j < n; j++) {
                    // when there is a queen at particular index I add Q to the
                    // StringBuilder or add '.'
                    if (box[i][j]) {
                        sb.append('Q');
                    } else {
                        sb.append('.');
                    }
                }
                li.add(sb.toString());
            }
            result.add(li);

        }
        // main logic

        for (int i = 0; i < n; i++) {
            // check is this place is safe to put queen there
            if (isSafe(box, r, i, n)) {
                // if yes I will make this index true
                // action
                box[r][i] = true;
                // recurse
                backTrack(box, r + 1, n);
                // backTrack
                box[r][i] = false;
            }
        }
    }

    private static boolean isSafe(boolean[][] box, int r, int c, int n) {
        // check column that is it any queen there in same column.
        for (int i = 0; i < r; i++) {
            if (box[i][c]) {
                return false;
            }
        }
        // check diagonally
        // to preserve original index
        int i = r;
        int j = c;
        while (i >= 0 && j >= 0) {
            if (box[i--][j--]) {
                return false;
            }
        }
        i = r;
        j = c;
        while (i >= 0 && j < n) {
            if (box[i--][j++]) {
                return false;
            }
        }
        // if queen is safe to place return true
        return true;

    }

    public static void main(String[] args) {
        int n = 4;
        System.out.println(solveNQueens(n));
    }
}

File: nqueen.go
// Time Complexity : O(n!)
// Space Complexity : O(n2)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
create a board to mark positions
call recursive with 0 index

in recursive
base case: if index is n
create string for each level by appending . for each 0 board value and Q for 1 value
append all to temp array and then to result

logic
for each row
check if current index and i are safe
if so set board at current position to 1
recursively call for next row
set board value to 0 again

is safe function check if any previous row same column is 1 then return false
check above diagonals have 1 then return false
*/
package main

import "fmt"

var res [][]string
var board [][]int
func solveNQueens(n int) [][]string {
	res = [][]string{}
	board = make([][]int, n)
	for i := 0; i<n;i++ {
		board[i] = make([]int, n)
	}

	h(n, 0)
	return res
}

func h(n int, index int) {
	if index == n {
		temp := []string{}
		for y:=0;y<n;y++ {
			s := ""
			for z :=0;z<n;z++ {
				if board[y][z] == 0 {
					s += "."
				} else {
					s += "Q"
				}
			}
			temp = append(temp, s)
		}
		res = append(res, temp)
		return
	}

	for i:=0; i<n;i++ {
		if isSafe(n, index, i) {
			board[index][i] = 1
			h(n, index + 1)
			board[index][i] = 0
		}
	}
}

func isSafe(n int, index int, i int) bool {
	for x := 0; x<= index;x++ {
		if board[x][i] == 1 {
			return false
		}
	}

	x := index
	y := i
	for x >= 0 && y >= 0 {
		if board[x][y] == 1 {
			return false
		}
		x--
		y--
	}
	x = index
	y = i
	for x >= 0 && y < n {
		if board[x][y] == 1 {
			return false
		}
		x--
		y++
	}
	return true
}

func MainNQueen() {
	fmt.Println(solveNQueens(4)) //expected [[".Q..","...Q","Q...","..Q."],["..Q.","Q...","...Q",".Q.."]]
}
